storage-proofs/src/circuit/rational_post.rs

```diff
@@ -0,0 +1,377 @@
+use std::marker::PhantomData;
+
+use bellperson::{Circuit, ConstraintSystem, SynthesisError};
+use fil_sapling_crypto::jubjub::JubjubEngine;
+use paired::bls12_381::{Bls12, Fr};
+
+use crate::circuit::por::{PoRCircuit, PoRCompound};
+use crate::circuit::variables::Root;
+use crate::compound_proof::{CircuitComponent, CompoundProof};
+use crate::drgraph;
+use crate::hasher::Hasher;
+use crate::merklepor;
+use crate::parameter_cache::{CacheableParameters, ParameterSetMetadata};
+use crate::proof::ProofScheme;
+use crate::rational_post::RationalPoSt;
+use crate::util::NODE_SIZE;
+
+/// This is the `RationalPoSt` circuit.
+pub struct RationalPoStCircuit<'a, E: JubjubEngine, H: Hasher> {
+    /// Paramters for the engine.
+    pub params: &'a E::Params,
+    pub commitments: Vec<Option<E::Fr>>,
+    pub leafs: Vec<Option<E::Fr>>,
+    #[allow(clippy::type_complexity)]
+    pub paths: Vec<Vec<Option<(E::Fr, bool)>>>,
+    _h: PhantomData<H>,
+}
+
+pub struct RationalPoStCompound<H>
+where
+    H: Hasher,
+{
+    _h: PhantomData<H>,
+}
+
+impl<E: JubjubEngine, C: Circuit<E>, P: ParameterSetMetadata, H: Hasher>
+    CacheableParameters<E, C, P> for RationalPoStCompound<H>
+{
+    fn cache_prefix() -> String {
+        String::from("proof-of-spacetime-rational")
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct ComponentPrivateInputs {}
+
+impl<'a, E: JubjubEngine, H: Hasher> CircuitComponent for RationalPoStCircuit<'a, E, H> {
+    type ComponentPrivateInputs = ComponentPrivateInputs;
+}
+
+impl<'a, H> CompoundProof<'a, Bls12, RationalPoSt<'a, H>, RationalPoStCircuit<'a, Bls12, H>>
+    for RationalPoStCompound<H>
+where
+    H: 'a + Hasher,
+{
+    fn generate_public_inputs(
+        pub_in: &<RationalPoSt<'a, H> as ProofScheme<'a>>::PublicInputs,
+        pub_params: &<RationalPoSt<'a, H> as ProofScheme<'a>>::PublicParams,
+        _partition_k: Option<usize>,
+    ) -> Vec<Fr> {
+        let mut inputs = Vec::new();
+
+        let por_pub_params = merklepor::PublicParams {
+            leaves: (pub_params.sector_size as usize / NODE_SIZE),
+            private: false,
+        };
+
+        for (challenge, commitment) in pub_in.challenges.iter().zip(pub_in.commitments) {
```

Comment:
SUGGESTION: I think it would be wise to assert that `pub_in.challenges` and `pub_in.commitments` have equal lengths.